### ISSUE TITLE
SMB shares fail to create SMB credentials file in Windows 10/CentOS7

### ISSUE THREAD

**Issue by dswitzer** [hashicorp/vagrant#11816]
I cannot get SMB file shares to work in CentOS 7. What appears to be happening in is the SMB credentials file is never written. It appears that the `cat` operation to write the /etc/smb_creds_* file never creates the file.

Here's what I'm seeing in the debug output:

```
 INFO ssh: Execute: cat <<"EOF" >/etc/smb_creds_vgt-33e1c4e78e60b788e3f4cd2af4ecddeb-bc759494af01ee8f2ca8d2ced5747fdf
username=myusername
password=*****

EOF
chmod 0600 /etc/smb_creds_vgt-33e1c4e78e60b788e3f4cd2af4ecddeb-bc759494af01ee8f2ca8d2ced5747fdf
 (sudo=true)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 0
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: mount -t cifs -o vers=2.0,credentials=/etc/smb_creds_vgt-33e1c4e78e60b788e3f4cd2af4ecddeb-bc759494af01ee8f2ca8d2ced5747fdf,uid=1001,gid=501,mfsymlinks,dmode=770,fmode=770 //192.168.1.168/vgt-33e1c4e78e60b788e3f4cd2af4ecddeb-bc759494af01ee8f2ca8d2ced5747fdf /var/www/site (sudo=true)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: stderr: mount error(22): Invalid argument
Refer to the mount.cifs(8) manual page (e.g. man mount.cifs)

DEBUG ssh: Exit status: 32
 INFO retryable: Retryable exception raised: #<Vagrant::Errors::LinuxMountFailed: Failed to mount folders in Linux guest. This is usually because
the "vboxsf" file system is not available. Please verify that
the guest additions are properly installed in the guest and
can work properly. The command attempted was:

mount -t cifs -o vers=2.0,credentials=/etc/smb_creds_vgt-33e1c4e78e60b788e3f4cd2af4ecddeb-bc759494af01ee8f2ca8d2ced5747fdf,uid=1001,gid=501,mfsymlinks,dmode=770,fmode=770 //192.168.1.168/vgt-33e1c4e78e60b788e3f4cd2af4ecddeb-bc759494af01ee8f2ca8d2ced5747fdf /var/www/site

The error output from the last command was:

mount error(22): Invalid argument
Refer to the mount.cifs(8) manual page (e.g. man mount.cifs)
```

### Vagrant version
Vagrant 2.2.9

### Host operating system

Windows 10 Home

### Guest operating system

CentOS 7

### Vagrantfile

```ruby
# -*- mode: ruby -*-
# vi: set ft=ruby :

Vagrant.configure(2) do |config|
  config.vm.box     = "private/centos-7-base"
  config.vm.box_url = "ftp://private.server.com/centos-7-base/package.box"

  config.vm.box_download_checksum_type="sha256"
  config.vm.box_download_checksum="8E1012AAA371E0AC3F3F0A7BB15D46C8E562553FFAA081B3FAF54B02CEC29D57"

  config.vm.network :forwarded_port, guest: 22, host: 2260, host_ip: "127.0.0.1", id: "ssh", auto_correct: true

  config.vm.hostname = "dev.local"
  config.vm.network "private_network", ip: "192.168.100.102"
  config.vm.network "private_network", ip: "192.168.100.103"
  config.vm.network "private_network", ip: "192.168.100.104"
  config.vm.network "private_network", ip: "192.168.100.105"

  config.vm.provider "virtualbox" do |vb|
    # Display the VirtualBox GUI when booting the machine
    vb.gui = false

    # Customize the amount of CPU & memory on the VM:
    vb.cpus = 2
    vb.memory = 4096
  end
  config.vm.synced_folder "./site/", "/var/www/site/", type: "smb", create: true, owner: 1001, group: 501, mount_options: ["dmode=770,fmode=770"]

end
```

### Debug output
[vagrant-debug.log](https://github.com/hashicorp/vagrant/files/5057428/vagrant-debug.log)

### Expected behavior
SMB file share should be created.

### Actual behavior

Vagrant fails to start, failing with the following error:

```
Failed to mount folders in Linux guest. This is usually because
the "vboxsf" file system is not available. Please verify that
the guest additions are properly installed in the guest and
can work properly. The command attempted was:

mount -t cifs -o vers=2.0,credentials=/etc/smb_creds_vgt-33e1c4e78e60b788e3f4cd2af4ecddeb-bc759494af01ee8f2ca8d2ced5747fdf,uid=1001,gid=501,mfsymlinks,dmode=770,fmode=770 //192.168.1.168/vgt-33e1c4e78e60b788e3f4cd2af4ecddeb-bc759494af01ee8f2ca8d2ced5747fdf /var/www/site

The error output from the last command was:

mount error(22): Invalid argument
Refer to the mount.cifs(8) manual page (e.g. man mount.cifs)
```

### Steps to reproduce

1. vagrant up

### References

I'm not sure this is the issue, but perhaps this is related to the following issue:

https://stackoverflow.com/questions/18836853/sudo-cat-eof-file-doesnt-work-sudo-su-does

**Comment by soapy1** [comment]
Hey there, 
thanks for opening an issue. The problem seems to be that the [`dmode` and `fmode` options are not available for a cifs mount](https://linux.die.net/man/8/mount.cifs). I was able to recreate this issue, and removing those two options resolved things.


> What appears to be happening in is the SMB credentials file is never written. It appears that the cat operation to write the /etc/smb_creds_* file never creates the file.

This behavior is because Vagrant will [delete this file](https://github.com/hashicorp/vagrant/blob/master/plugins/guests/linux/cap/mount_smb_shared_folder.rb#L85) after the mount occurs (or fails) so that there aren't a set of root credentials floating around.

Cheers!

**Comment by dswitzer** [comment]
@soapy1

You are indeed correct! 

I discovered this just after posting this issue and was coming back to update the issue. The file was getting created, but just deleted before I was able to see it. So I thought the issue was that it couldn't find the SMB credentials file. It didn't occur to me that the `dmode` and `fmode` options were invalid.

Sorry I didn't catch this sooner. Maybe this issue will help someone else having the same problem in the future.
